compiler/src/test/scala/org/scalajs/core/compiler/test/JSOptionalTest.scala

```diff
@@ -0,0 +1,141 @@
+      |newSource1.scala:10: error: Members of Scala.js-defined JS traits must either be abstract, or their right-hand-side must be `js.undefined`.
+      |      def b1: js.UndefOr[Int] = 5
+      |                                ^
+      |newSource1.scala:11: error: Members of Scala.js-defined JS traits must either be abstract, or their right-hand-side must be `js.undefined`.
+      |      def b2: Int = 5
+      |                    ^
+      |newSource1.scala:13: error: Members of Scala.js-defined JS traits must either be abstract, or their right-hand-side must be `js.undefined`.
+      |      var c1: js.UndefOr[Int] = 5
+      |                                ^
+      |newSource1.scala:14: error: Members of Scala.js-defined JS traits must either be abstract, or their right-hand-side must be `js.undefined`.
+      |      var c2: Int = 5
+      |                    ^
+    """
+  }
+
+  @Test
+  def noOverrideNonOptionalWithOptional: Unit = {
+    s"""
+    @ScalaJSDefined
+    abstract class A extends js.Object {
+      val a1: js.UndefOr[Int] = 5
+      val a2: js.UndefOr[Int]
+
+      def b1: js.UndefOr[Int] = 5
+      def b2: js.UndefOr[Int]
+    }
+
+    @ScalaJSDefined
+    trait B extends A {
+      override val a1: js.UndefOr[Int] = js.undefined
+      override val a2: js.UndefOr[Int] = js.undefined
+
+      override def b1: js.UndefOr[Int] = js.undefined
+      override def b2: js.UndefOr[Int] = js.undefined
+    }
+    """ hasErrors
+    s"""
+      |newSource1.scala:16: error: Cannot override val a1: scala.scalajs.js.UndefOr[Int] with override val a1: scala.scalajs.js.UndefOr[Int], because the latter is concrete and declared in a JavaScript trait.
+      |      override val a1: js.UndefOr[Int] = js.undefined
+      |                   ^
+      |newSource1.scala:17: error: Cannot override val a2: scala.scalajs.js.UndefOr[Int] with override val a2: scala.scalajs.js.UndefOr[Int], because the latter is concrete and declared in a JavaScript trait.
+      |      override val a2: js.UndefOr[Int] = js.undefined
+      |                   ^
+      |newSource1.scala:19: error: Cannot override def b1: scala.scalajs.js.UndefOr[Int] with override def b1: scala.scalajs.js.UndefOr[Int], because the latter is concrete and declared in a JavaScript trait.
+      |      override def b1: js.UndefOr[Int] = js.undefined
+      |                   ^
+      |newSource1.scala:20: error: Cannot override def b2: scala.scalajs.js.UndefOr[Int] with override def b2: scala.scalajs.js.UndefOr[Int], because the latter is concrete and declared in a JavaScript trait.
+      |      override def b2: js.UndefOr[Int] = js.undefined
+      |                   ^
+    """
+
+    s"""
+    @ScalaJSDefined
+    trait A extends js.Object {
+      val a: js.UndefOr[Int]
+      def b: js.UndefOr[Int]
+    }
+
+    @ScalaJSDefined
+    trait B extends A {
+      override val a: js.UndefOr[Int] = js.undefined
+      override def b: js.UndefOr[Int] = js.undefined
```

Comment:
I do not think there is a good reason. I was planning to try and remove that restriction.